src/Step/Deploy/InstallUpdate/ConfigUpdate/DbConnection.php

```diff
@@ -60,83 +64,321 @@ class DbConnection implements StepInterface
+            $this->logger->notice(
+                'Database was already split but deploy was configured with new connection.'
+                . ' The previous connection data will be ignored.'
+            );
+        }
+
+        $useSlave = $this->stageConfig->get(DeployInterface::VAR_MYSQL_USE_SLAVE_CONNECTION);
+        $slaveIsAvailable = isset($dbConfig[DbConfig::KEY_SLAVE_CONNECTION]);
+
+        if (empty($mageSplitConnectionsConfig)
+            || (!empty($mageSplitConnectionsConfig) && $isCustomDefaultConnection)) {
+            $this->updateMainConnectionsConfig($useSlave, $slaveIsAvailable);
+            return;
+        }
+
+        $customSplitConnections = $this->getDifferentConnections(
+            $mageSplitConnectionsConfig,
+            $dbConfig[DbConfig::KEY_CONNECTION]
+        );
+
+        if (!empty($customSplitConnections)) {
+            $this->logger->warning(sprintf(
+                'For split databases used custom connections: %s',
+                implode(', ', $customSplitConnections)
+            ));
+            $this->flagManager->set(FlagManager::FLAG_IGNORE_SPLIT_DB);
+            return;
+        }
+
+        $this->updateSlaveConnectionsConfig($useSlave, $slaveIsAvailable);
     }
 
     /**
-     * Adds logging about slave connection.
+     * Update main connection configurations of app/etc/env.php
+     *
+     * @param bool $useSlave
+     * @param bool $slaveIsAvailable
+     * @throws FileSystemException
+     * @throws ConfigException
      */
-    private function addLoggingAboutSlaveConnection()
+    public function updateMainConnectionsConfig(
+        bool $useSlave,
+        bool $slaveIsAvailable
+    ) {
+        $mageConfig = $this->getMageConfigData();
+        $mageConfig[DbConfig::KEY_DB] = $this->getMainDbConfig($useSlave && $slaveIsAvailable);
+        $mageConfig[ResourceConfig::KEY_RESOURCE] = $this->getMainResourceConfig();
+        $this->addLoggingAboutSlaveConnection($mageConfig[DbConfig::KEY_DB], $useSlave);
+        $this->configWriter->create($mageConfig);
+    }
+
+    /**
+     * Updates db slave configurations
+     *
+     * @param bool $useSlave
+     * @param bool $slaveIsAvailable
+     * @throws FileSystemException
+     * @throws ConfigException
+     */
+    private function updateSlaveConnectionsConfig(
+        bool $useSlave,
+        bool $slaveIsAvailable
+    ) {
+        $mageConfig = $this->getMageConfigData();
+        if ($useSlave && $slaveIsAvailable) {
+            $dbConfig = $this->getDbConfigData();
+            $slaveConnectionsConfig = $this->getMainConnections($dbConfig[DbConfig::KEY_SLAVE_CONNECTION]);
+            $mageConfig[DbConfig::KEY_DB][DbConfig::KEY_SLAVE_CONNECTION] = $slaveConnectionsConfig;
+        } else {
+            unset($mageConfig[DbConfig::KEY_DB][DbConfig::KEY_SLAVE_CONNECTION]);
+        }
+        $this->addLoggingAboutSlaveConnection($mageConfig[DbConfig::KEY_DB], $useSlave);
```

Comment:
You should move this line into if statement `if ($useSlave && $slaveIsAvailable) {`
For example, if $useSlave is true and $slaveIsAvailable is false slave configuration won't update but logs will be added.

Reply:
Done